src/wasm/wasm-type.cpp

```diff
@@ -104,38 +172,79 @@ void Type::init(const std::vector<Type>& types) {
     *this = types[0];
     return;
   }
+  id = canonicalize(TypeDef(tuple));
+}
 
-  // Add a new type if it hasn't been added concurrently
-  std::lock_guard<std::mutex> lock(mutex);
-  auto indexIt = indices.find(types);
-  if (indexIt != indices.end()) {
-    id = indexIt->second;
+Type::Type(const Signature signature, bool nullable) {
+  id = canonicalize(TypeDef(signature, nullable));
+}
+
+Type::Type(const Struct& struct_, bool nullable) {
+#ifndef NDEBUG
+  for (Field f : struct_.fields) {
+    assert(f.type.isSingle() && f.type.isConcrete());
+  }
+#endif
+  id = canonicalize(TypeDef(struct_, nullable));
+}
+
+Type::Type(const Array& array, bool nullable) {
+#ifndef NDEBUG
+  assert(array.element.type.isSingle() && array.element.type.isConcrete());
+#endif
```

Comment:
```suggestion
  assert(array.element.type.isSingle());
```
No need for the `NDEBUG` check since `assert` already does it.